rustler_mix/lib/rustler/compiler.ex

```diff
@@ -69,46 +70,58 @@ defmodule Rustler.Compiler do
     ["rustup", "run", version, "cargo", "rustc"]
   end
 
-  defp make_platform_hacks(args, crate_path, {:unix, :darwin}) do
-    root = Path.join([".cargo", "config"])
-    path = Path.join([crate_path, ".cargo", "config"])
-
-    if File.exists?(root) || File.exists?(path) do
-      args
-    else
-      IO.write([
-        "\n",
-        IO.ANSI.yellow(),
-        """
-        Compiling on macOS requires special link args in order to compile
-        correctly.
-
-        Rustler is currently working around this issue in the compiler task.
-        This will be removed in v1.0.0 in favor of a user supplied .cargo/config
-        file.
-
-        To remove this warning, please create #{path}
-        with the following content:
-
-              [target.x86_64-apple-darwin]
-              rustflags = [
-                  "-C", "link-arg=-undefined",
-                  "-C", "link-arg=dynamic_lookup",
-              ]
-
-        See https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/MachOTopics/1-Articles/executing_files.html
-        for more details.
-
-        """,
-        IO.ANSI.default_color(),
-        "\n"
-      ])
-
-      args ++ ["--", "-C", "link-arg=-undefined", "-C", "link-arg=dynamic_lookup"]
+  defp ensure_platform_requirements!(crate_path, config, {:unix, :darwin}) do
+    workspace_root = config.metadata["workspace_root"]
+
+    components =
+      crate_path
+      |> Path.relative_to(workspace_root)
+      |> Path.split()
+
+    {potential_config_files, _} =
+      Enum.map_reduce(["" | components], workspace_root, fn component, path ->
+        path = Path.join(path, component)
+        file = Path.join([path, ".cargo", "config"])
+        {file, path}
+      end)
+
+    sorted_config_files =
+      potential_config_files
+      |> Enum.filter(&File.exists?/1)
+      |> Enum.reverse()
+
+    has_macos_target_os_configuration? =
+      sorted_config_files
+      |> Stream.map(&Toml.decode_file!/1)
+      |> Enum.drop_while(fn toml ->
+        toml
+        |> Map.get("target", [])
+        |> Enum.filter(fn {key, _} ->
+          String.match?(key, ~r/(.*macos.*)|(.*darwin.*)/)
```

Comment:
Couldn't we check the args here as suggested by @filmor?

```suggestion
          String.match?(key, ~r/(.*macos.*)|(.*darwin.*)/) and String.match?(value, ~r/(.*link-arg=-undefined.*)|(.*link-arg=dynamic_lookup.*)/)
```

I'm not sure if `value` is a string, though.

Reply:
Ah, maybe I misunderstood @filmor, I thought he suggested to add the correct arguments to `args`. I think we can check the correct arguments here. I will figure out what the Toml parser returns for the key-value pairs.

Reply:
See f16cbf2..2697b62, I added some code to match the specific arguments. I think this might be a bit brittle. Maybe it is enough if the target configuration exists at all, as that implies that a developer already thought about what the correct arguments should be?

Reply:
Yeah, I think it would be enough to check if the target config exists :+1: